ansq/tcp/types/connection.py

```diff
@@ -21,73 +50,114 @@ def __init__(
-        tls_v1: bool = False,
-        snappy: bool = False,
-        deflate: bool = False,
-        deflate_level: int = 6,
-        sample_rate: int = 0,
-        debug: bool = False,
-        logger: logging.Logger = None,
+        connection_options: ConnectionOptions = ConnectionOptions(),
+        message_queue: Optional[asyncio.Queue] = None,
+        on_message: Optional[Callable] = None,
+        on_exception: Optional[Callable] = None,
+        loop: Optional[AbstractEventLoop] = None,
+        auto_reconnect: Optional[bool] = None,
+        deflate: Optional[bool] = None,
+        deflate_level: Optional[int] = None,
+        tls_v1: Optional[bool] = None,
+        snappy: Optional[bool] = None,
+        sample_rate: Optional[int] = None,
+        heartbeat_interval: Optional[int] = None,
+        feature_negotiation: Optional[bool] = None,
+        debug: Optional[bool] = None,
+        logger: Optional[logging.Logger] = None,
     ):
         from ansq.tcp.protocol import Reader
         from ansq.tcp.types import ConnectionStatus
         from ansq.utils import get_logger
 
+        deprecated_kw_options: Dict[str, Any] = {
+            "message_queue": message_queue,
+            "on_message": on_message,
+            "on_exception": on_exception,
+            "loop": loop,
+            "auto_reconnect": auto_reconnect,
+            "debug": debug,
+            "logger": logger,
+        }
+        deprecated_kw_features: Dict[str, Any] = {
+            "deflate": deflate,
+            "deflate_level": deflate_level,
+            "feature_negotiation": feature_negotiation,
+            "heartbeat_interval": heartbeat_interval,
+            "sample_rate": sample_rate,
+            "snappy": snappy,
+            "tls_v1": tls_v1,
+        }
+        if any(
+            arg is not None
+            for arg in itertools.chain(
+                deprecated_kw_options.values(), deprecated_kw_features.values()
+            )
+        ):
+            warnings.warn(
+                message=(
+                    "Passing connection options to `TCPConnection` using keyword "
+                    "arguments is deprecated: use `ConnectionOptions` structure instead"
+                ),
+                category=DeprecationWarning,
+            )
+            connection_features = attr.evolve(
+                ConnectionFeatures(),
+                **{
+                    name: value
+                    for name, value in deprecated_kw_features.items()
+                    if value is not None
+                },
+            )
+            deprecated_kw_options["features"] = connection_features
+            connection_options = attr.evolve(
+                connection_options,
+                **{
+                    name: value
+                    for name, value in deprecated_kw_options.items()
+                    if value is not None
+                },
+            )
+        self._options: ConnectionOptions = connection_options
```

Comment:
I have an idea to simplify this a little bit. See https://github.com/lig/ansq/pull/1. Looks similar to `conn.identify()` method. What do you think?

Reply:
I like the idea. No reason there in keeping deperecated kwargs defined explicitly. This way, a user will be gently forced to use new data structure in case they're modifying the code relying on deprecated kwargs.

Reply:
Merged https://github.com/lig/ansq/pull/1